django_ontruck/denormalisation.py

```diff
@@ -0,0 +1,350 @@
+    """
+    def __init__(
+        self,
+        model_class,
+        prefetch_related_objects_=prefetch_related_objects,
+        fields=None
+    ):
+        self.model_class = model_class
+        self._fields = fields
+        self.prefetch_related_objects = prefetch_related_objects_
+
+    def limit_fields(self, fields):
+        self._fields = fields
+
+        return self
+
+    @property
+    def all_fields(self):
+        return {prop.name for prop in self.model_class.denormalised_properties}
+
+    @property
+    def fields(self):
+        if self._fields is not None:
+            return set(self._fields) & self.all_fields
+
+        return self.all_fields
+
+    @property
+    def prefetches(self):
+        return flatten(
+            prop.prefetches
+            for prop in self.model_class.denormalised_properties
+        )
+
+    @property
+    def parent_model_class(self):
+        return self.model_class.parent.field.related_model
+
+    @property
+    def app_label(self):
+        return self.model_class._meta.app_label  # noqa
+
+    def __call__(self, sender, **kwargs):
+        instance = kwargs['instance']
+
+        self.prefetch_related_objects([instance], *self.prefetches)
+
+        self.model_class.objects.update_or_create(
+            parent_id=instance.id,
+            defaults={
+                field: getattr(instance, field) for field in self.fields
+            }
+        )
+
+
+class DenormaliserSerialiser:
+    """
+    Dumps a Denormaliser to JSON serialisable dict.  This is allows the
+    denormaliser to be sent over the wire and invoked in a background task.
+    """
+    def __init__(self, denormaliser):
+        self.denormaliser = denormaliser
+
+    def dump(self):
+        return {
+            'model_class_module': self.denormaliser.model_class.__module__,
+            'model_class_name': self.denormaliser.model_class.__name__,
+            'fields': [*self.denormaliser.fields]
+        }
+
+
+class DenormaliserDeserialiser:
+    """
+    Loads a Denormaliser from a JSON serialisable dict.
+    """
+    def __init__(self, data, import_module_=import_module):
+        self.model_class_module = data['model_class_module']
+        self.model_class_name = data['model_class_name']
+        self.fields = data['fields']
+
+        self.import_module = import_module_
+
+    @property
+    def module(self):
+        return self.import_module(self.model_class_module)
+
+    @property
+    def klass(self):
+        return getattr(self.module, self.model_class_name)
+
+    def load(self):
+        return Denormaliser(model_class=self.klass, fields=self.fields)
+
+
+def trivial_decorator(func):
+    return func
+
+
+class DecoratedSignal:
+    """
+    Adapt any signal to the interface of Denormalise via a decorator.
+
+    :param signal:  The signal we wish to handle
+    :param decorator: A decorator which, when passed a Denormaliser instance,
```

Comment:
I don't find the name but I `decorator` for me is a bit generic.

Reply:
Completely agree.